tools/autograd/templates/variable_factories.h

```diff
@@ -48,7 +48,11 @@ AT_FORALL_SCALAR_TYPES_AND3(Bool, Half, BFloat16, TENSOR)
         sizes_(),
         scalar_type_(),
         type_(InitListTensorType::InitList) {
-      TORCH_CHECK(init_list.size() > 0, "Empty init-list is not supported");
+      TORCH_CHECK(
+        init_list.size() > 0,
+        "Empty init-list is not yet supported. We can create tensors with zero-size dimensions in the following way:\n",
```

Comment:
I know I asked you to add this error message, but I'm wondering: what actually happens if a user tries to call `torch::tensor({})`? Is it a compilation error or does this error message get triggered?

If this error message gets triggered, could we add a test for it?

Reply:
Compiling `torch::tensor({})` gives an "ambiguous overload" compilation error:
```
../test/cpp/api/tensor.cpp:232:35: error: call of overloaded ‘tensor(<brace-enclosed initializer list>)’ is ambiguous
     auto tensor = torch::tensor({});
                                   ^
In file included from ../caffe2/../torch/csrc/api/include/torch/types.h:7:0,
                 from ../caffe2/../torch/csrc/api/include/torch/detail/static.h:4,
                 from ../caffe2/../torch/csrc/api/include/torch/nn/pimpl.h:4,
                 from ../caffe2/../torch/csrc/api/include/torch/nn/module.h:3,
                 from ../caffe2/../torch/csrc/api/include/torch/nn/cloneable.h:3,
                 from ../test/cpp/api/support.h:7,
                 from ../test/cpp/api/tensor.cpp:2:
../torch/csrc/autograd/generated/variable_factories.h:181:644: note: candidate: at::Tensor torch::tensor(c10::ArrayRef<unsigned char>)
../torch/csrc/autograd/generated/variable_factories.h:181:766: note: candidate: at::Tensor torch::tensor(std::initializer_list<unsigned char>)
../torch/csrc/autograd/generated/variable_factories.h:181:887: note: candidate: at::Tensor torch::tensor(uint8_t)
../torch/csrc/autograd/generated/variable_factories.h:181:1603: note: candidate: at::Tensor torch::tensor(c10::ArrayRef<signed char>)
../torch/csrc/autograd/generated/variable_factories.h:181:1724: note: candidate: at::Tensor torch::tensor(std::initializer_list<signed char>)
../torch/csrc/autograd/generated/variable_factories.h:181:1843: note: candidate: at::Tensor torch::tensor(int8_t)
../torch/csrc/autograd/generated/variable_factories.h:181:2562: note: candidate: at::Tensor torch::tensor(c10::ArrayRef<short int>)
../torch/csrc/autograd/generated/variable_factories.h:181:2685: note: candidate: at::Tensor torch::tensor(std::initializer_list<short int>)
../torch/csrc/autograd/generated/variable_factories.h:181:2806: note: candidate: at::Tensor torch::tensor(int16_t)
../torch/csrc/autograd/generated/variable_factories.h:181:3507: note: candidate: at::Tensor torch::tensor(c10::ArrayRef<int>)
../torch/csrc/autograd/generated/variable_factories.h:181:3624: note: candidate: at::Tensor torch::tensor(std::initializer_list<int>)
../torch/csrc/autograd/generated/variable_factories.h:181:3737: note: candidate: at::Tensor torch::tensor(int)
../torch/csrc/autograd/generated/variable_factories.h:181:4450: note: candidate: at::Tensor torch::tensor(c10::ArrayRef<long int>)
../torch/csrc/autograd/generated/variable_factories.h:181:4572: note: candidate: at::Tensor torch::tensor(std::initializer_list<long int>)
../torch/csrc/autograd/generated/variable_factories.h:181:4693: note: candidate: at::Tensor torch::tensor(int64_t)
../torch/csrc/autograd/generated/variable_factories.h:181:5404: note: candidate: at::Tensor torch::tensor(c10::ArrayRef<float>)
../torch/csrc/autograd/generated/variable_factories.h:181:5525: note: candidate: at::Tensor torch::tensor(std::initializer_list<float>)
../torch/csrc/autograd/generated/variable_factories.h:181:5642: note: candidate: at::Tensor torch::tensor(float)
../torch/csrc/autograd/generated/variable_factories.h:181:6354: note: candidate: at::Tensor torch::tensor(c10::ArrayRef<double>)
../torch/csrc/autograd/generated/variable_factories.h:181:6477: note: candidate: at::Tensor torch::tensor(std::initializer_list<double>)
../torch/csrc/autograd/generated/variable_factories.h:181:6596: note: candidate: at::Tensor torch::tensor(double)
../torch/csrc/autograd/generated/variable_factories.h:181:7630: note: candidate: at::Tensor torch::tensor(c10::ArrayRef<bool>)
../torch/csrc/autograd/generated/variable_factories.h:181:7815: note: candidate: at::Tensor torch::tensor(std::initializer_list<bool>)
../torch/csrc/autograd/generated/variable_factories.h:181:8062: note: candidate: at::Tensor torch::tensor(c10::impl::ScalarTypeToCPPType<(c10::ScalarType)11>::type)
../torch/csrc/autograd/generated/variable_factories.h:181:9224: note: candidate: at::Tensor torch::tensor(c10::ArrayRef<c10::Half>)
../torch/csrc/autograd/generated/variable_factories.h:181:9409: note: candidate: at::Tensor torch::tensor(std::initializer_list<c10::Half>)
../torch/csrc/autograd/generated/variable_factories.h:181:9656: note: candidate: at::Tensor torch::tensor(c10::impl::ScalarTypeToCPPType<(c10::ScalarType)5>::type)
../torch/csrc/autograd/generated/variable_factories.h:181:10838: note: candidate: at::Tensor torch::tensor(c10::ArrayRef<c10::BFloat16>)
../torch/csrc/autograd/generated/variable_factories.h:181:11031: note: candidate: at::Tensor torch::tensor(std::initializer_list<c10::BFloat16>)
../torch/csrc/autograd/generated/variable_factories.h:181:11286: note: candidate: at::Tensor torch::tensor(c10::impl::ScalarTypeToCPPType<(c10::ScalarType)15>::type)
In file included from ../caffe2/../torch/csrc/api/include/torch/types.h:7:0,
                 from ../caffe2/../torch/csrc/api/include/torch/detail/static.h:4,
                 from ../caffe2/../torch/csrc/api/include/torch/nn/pimpl.h:4,
                 from ../caffe2/../torch/csrc/api/include/torch/nn/module.h:3,
                 from ../caffe2/../torch/csrc/api/include/torch/nn/cloneable.h:3,
                 from ../test/cpp/api/support.h:7,
                 from ../test/cpp/api/tensor.cpp:2:
../torch/csrc/autograd/generated/variable_factories.h:197:19: note: candidate: at::Tensor torch::tensor(torch::detail::InitListTensor)
 inline at::Tensor tensor(detail::InitListTensor init_list_tensor) {
                   ^
In file included from ../caffe2/../torch/csrc/api/include/torch/types.h:7:0,
                 from ../caffe2/../torch/csrc/api/include/torch/detail/static.h:4,
                 from ../caffe2/../torch/csrc/api/include/torch/nn/pimpl.h:4,
                 from ../caffe2/../torch/csrc/api/include/torch/nn/module.h:3,
                 from ../caffe2/../torch/csrc/api/include/torch/nn/cloneable.h:3,
                 from ../test/cpp/api/support.h:7,
                 from ../test/cpp/api/tensor.cpp:2:
../torch/csrc/autograd/generated/variable_factories.h:212:19: note: candidate: at::Tensor torch::tensor(std::initializer_list<torch::detail::InitListTensor>)
 inline at::Tensor tensor(std::initializer_list<detail::InitListTensor> init_list) {
                   ^
```
Compiling `torch::tensor({{}, {}})` also gives an "ambiguous overload" compilation error:
```
../test/cpp/api/tensor.cpp:232:41: error: call of overloaded ‘tensor(<brace-enclosed initializer list>)’ is ambiguous
     auto tensor = torch::tensor({{}, {}});
                                         ^
In file included from ../caffe2/../torch/csrc/api/include/torch/types.h:7:0,
                 from ../caffe2/../torch/csrc/api/include/torch/detail/static.h:4,
                 from ../caffe2/../torch/csrc/api/include/torch/nn/pimpl.h:4,
                 from ../caffe2/../torch/csrc/api/include/torch/nn/module.h:3,
                 from ../caffe2/../torch/csrc/api/include/torch/nn/cloneable.h:3,
                 from ../test/cpp/api/support.h:7,
                 from ../test/cpp/api/tensor.cpp:2:
../torch/csrc/autograd/generated/variable_factories.h:181:644: note: candidate: at::Tensor torch::tensor(c10::ArrayRef<unsigned char>)
../torch/csrc/autograd/generated/variable_factories.h:181:766: note: candidate: at::Tensor torch::tensor(std::initializer_list<unsigned char>)
../torch/csrc/autograd/generated/variable_factories.h:181:1603: note: candidate: at::Tensor torch::tensor(c10::ArrayRef<signed char>)
../torch/csrc/autograd/generated/variable_factories.h:181:1724: note: candidate: at::Tensor torch::tensor(std::initializer_list<signed char>)
../torch/csrc/autograd/generated/variable_factories.h:181:2562: note: candidate: at::Tensor torch::tensor(c10::ArrayRef<short int>)
../torch/csrc/autograd/generated/variable_factories.h:181:2685: note: candidate: at::Tensor torch::tensor(std::initializer_list<short int>)
../torch/csrc/autograd/generated/variable_factories.h:181:3507: note: candidate: at::Tensor torch::tensor(c10::ArrayRef<int>)
../torch/csrc/autograd/generated/variable_factories.h:181:3624: note: candidate: at::Tensor torch::tensor(std::initializer_list<int>)
../torch/csrc/autograd/generated/variable_factories.h:181:4450: note: candidate: at::Tensor torch::tensor(c10::ArrayRef<long int>)
../torch/csrc/autograd/generated/variable_factories.h:181:4572: note: candidate: at::Tensor torch::tensor(std::initializer_list<long int>)
../torch/csrc/autograd/generated/variable_factories.h:181:5404: note: candidate: at::Tensor torch::tensor(c10::ArrayRef<float>)
../torch/csrc/autograd/generated/variable_factories.h:181:5525: note: candidate: at::Tensor torch::tensor(std::initializer_list<float>)
../torch/csrc/autograd/generated/variable_factories.h:181:6354: note: candidate: at::Tensor torch::tensor(c10::ArrayRef<double>)
../torch/csrc/autograd/generated/variable_factories.h:181:6477: note: candidate: at::Tensor torch::tensor(std::initializer_list<double>)
../torch/csrc/autograd/generated/variable_factories.h:181:7630: note: candidate: at::Tensor torch::tensor(c10::ArrayRef<bool>)
../torch/csrc/autograd/generated/variable_factories.h:181:7815: note: candidate: at::Tensor torch::tensor(std::initializer_list<bool>)
../torch/csrc/autograd/generated/variable_factories.h:181:9224: note: candidate: at::Tensor torch::tensor(c10::ArrayRef<c10::Half>)
../torch/csrc/autograd/generated/variable_factories.h:181:9409: note: candidate: at::Tensor torch::tensor(std::initializer_list<c10::Half>)
../torch/csrc/autograd/generated/variable_factories.h:181:9656: note: candidate: at::Tensor torch::tensor(c10::impl::ScalarTypeToCPPType<(c10::ScalarType)5>::type)
../torch/csrc/autograd/generated/variable_factories.h:181:10838: note: candidate: at::Tensor torch::tensor(c10::ArrayRef<c10::BFloat16>)
../torch/csrc/autograd/generated/variable_factories.h:181:11031: note: candidate: at::Tensor torch::tensor(std::initializer_list<c10::BFloat16>)
../torch/csrc/autograd/generated/variable_factories.h:181:11286: note: candidate: at::Tensor torch::tensor(c10::impl::ScalarTypeToCPPType<(c10::ScalarType)15>::type)
In file included from ../caffe2/../torch/csrc/api/include/torch/types.h:7:0,
                 from ../caffe2/../torch/csrc/api/include/torch/detail/static.h:4,
                 from ../caffe2/../torch/csrc/api/include/torch/nn/pimpl.h:4,
                 from ../caffe2/../torch/csrc/api/include/torch/nn/module.h:3,
                 from ../caffe2/../torch/csrc/api/include/torch/nn/cloneable.h:3,
                 from ../test/cpp/api/support.h:7,
                 from ../test/cpp/api/tensor.cpp:2:
../torch/csrc/autograd/generated/variable_factories.h:197:19: note: candidate: at::Tensor torch::tensor(torch::detail::InitListTensor)
 inline at::Tensor tensor(detail::InitListTensor init_list_tensor) {
                   ^
In file included from ../caffe2/../torch/csrc/api/include/torch/types.h:7:0,
                 from ../caffe2/../torch/csrc/api/include/torch/detail/static.h:4,
                 from ../caffe2/../torch/csrc/api/include/torch/nn/pimpl.h:4,
                 from ../caffe2/../torch/csrc/api/include/torch/nn/module.h:3,
                 from ../caffe2/../torch/csrc/api/include/torch/nn/cloneable.h:3,
                 from ../test/cpp/api/support.h:7,
                 from ../test/cpp/api/tensor.cpp:2:
../torch/csrc/autograd/generated/variable_factories.h:212:19: note: candidate: at::Tensor torch::tensor(std::initializer_list<torch::detail::InitListTensor>)
 inline at::Tensor tensor(std::initializer_list<detail::InitListTensor> init_list) {
                   ^
```
I think improving the error message here is still a good idea, in case people try to look at the source code to figure out what they should do to create tensors with zero-size dimensions.

Reply:
Is it clearly documented somewhere that we don't support `{}`?

Reply:
Currently it's documented in https://github.com/pytorch/pytorch/pull/26890/files#diff-11228124349c21669d7262f207725b41R184-R188. I will also update https://pytorch.org/cppdocs/notes/tensor_creation.html to document this, after merging this PR.